### ISSUE TITLE
OOM error; mem cost keep growing!

### ISSUE THREAD

**Issue by 35p32** [rabbityl/lepard#5]
(Thanks for your excellent work bro !  this paper is really nice and amazing~
When i try to  train the  lepard myself ( I only change  the "batchsize=1"  and "num_workers=0"),  i found the memory  is keep raising(i have 126g mem,  lepard almost cost 100g and still raising ),  finally  it was killed by linux

Have you encountered this kind of problem before? 
 I did not change your code,   i guess this problem come from dataloader?  can you give me some suggestions ?

**Comment by rabbityl** [comment]
Hi, I just tested this code again with bsize=8 on a 80GB GPU card. I did not got a OOM error. 
Since you have a much larger GPU mem. It should be fine.
If its the dataloader, you can try to modify the dataset to over-fit only one sample.

**Comment by 35p32** [comment]
thanks for your replay , i have enough GPU ,  the problem is the "memory” OOM
when i just loop the dataloader and do nothing else, i found the "memory"    keep growing. 
May be memory leakage?

**Comment by 35p32** [comment]
For example, in your "forward" source code ,   i  use memory moniter to debug the memory leckage.  I  find some problem blew
(Right part is your source code， left part is the memory cost（not gpu cost）for every line, please mention the "Increment " colume which means the increase memory for every forward
![bad](https://user-images.githubusercontent.com/34885425/147339921-43c48fcd-a931-462f-a52d-cddf913b26a8.png)

When i forward once  , [Line 26] increase 3.9M memory ， when i forward twice ,[Line 26] may be increase  3.9 + 3.9 =7.8M memory.  When i forward one epoch(about 20000+ times) , [Line 26] increase 80G memory

**Comment by Yang-L1** [comment]
Does this mean all the functions causes memory leakage? 
The training takes 15+ epochs on my machine, and works totally fine.
No idea of what's going on.

**Comment by 35p32** [comment]
> 



> Does this mean all the functions causes memory leakage? The training takes 15+ epochs on my machine, and works totally fine. No idea of what's going on.

Sir,Could you please let me know if you can reproduce the  result  on 3DMatch&3DLomatch?

**Comment by 35p32** [comment]
> Does this mean all the functions causes memory leakage? The training takes 15+ epochs on my machine, and works totally fine. No idea of what's going on.

Thank you so much

**Comment by Yang-L1** [comment]
> > 
> 
> > Does this mean all the functions causes memory leakage? The training takes 15+ epochs on my machine, and works totally fine. No idea of what's going on.
> 
> Sir,Could you please let me know if you can reproduce the result on 3DMatch&3DLomatch?

Yes, I can reproduce the results.

**Comment by chalth** [comment]
Hi， I meet the same problem,  did you slove it?

**Comment by ACuOoOoO** [comment]
I find that the problem is caused by AverageMeter.update() in a training step. I solve the problem by detach the input tensor during the accumulation in AverageMeter.update().
